### PR TITLE
[enh] Adding optionnal recovery password + delete interface using password

### PR DESCRIPTION
**Pull request by alexAubin**
This is an attempt to implement [this feature](https://dev.yunohost.org/issues/6)

- When registering a domain, user can provide an optionnal field `recovery_password`. This is *optionnal* so this is *transparent and backward-compatible* for previous client implementations.
- `recovery_password` can be use to delete the domain (e.g. to be able to re-register it). A web-interface is provided at '/delete' (see delete.html) where user can input the domain name and the recovery password. No fancy CSS for now, just raw input boxes.
- On the server side, `recovery_password` is hashed and salted using BCrypt. Should be the standard security scheme.
- On the client side, `recovery_password` is generated using a hash of the user-inputted "true" password, with the domain as a semi-static salt : `recovery_password = sha256(sub.domain.tld:true_password)[0:32]`. The intention behind this is only to never send the user-inputted password, and protect against password reuse, in case of communication interception (by a third party, or because dynette server go compromised by the [FVEY](https://fr.wikipedia.org/wiki/Five_Eyes) and can listen to incoming password before they are even processed/stored)(Tin foil hat, yes 😛).
- This require a small migration on current dynette's server to work, more precisely to add a `recovery_password` column in the existing db. I'm really noob with ruby and databases, but hopefully this shouldn't be too difficult.
- Another PR is need on `yunohost/dyndns.py` to ask the user for a password and add `recovery_password` in the domain registering request.

In case somebody wants to test it, here are some quick-and-dirty tests I wrote, [here](https://paste.yunohost.org/ilibakumez.coffee).

### CODE REVIEW

**Comment by zamentur** [comment]
why don't you use the public key directly ?

**Comment by alexAubin** [comment]
Not sure I get the question, but the public/private keys are still used like before (e.g. for updating the domain). The `recovery_password` is only used when a user want to remove "by hand" a given domain, and we can't use the public/private key because the point is that the user should remember it :smile:.

File: delete.html [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+ * @param   {Object} [options]
+ * @param   {string} [options.msgFormat=string] - Message format: 'string' for JavaScript string
+ *   (gets converted to UTF-8 for hashing); 'hex-bytes' for string of hex bytes ('616263' ≡ 'abc') .
+ * @param   {string} [options.outFormat=hex] - Output format: 'hex' for string of contiguous
+ *   hex bytes; 'hex-w' for grouping hex bytes into groups of (4 byte / 8 character) words.
+ * @returns {string} Hash of msg as hex character string.
+ */
+Sha256.hash = function(msg, options) {
+    var defaults = { msgFormat: 'string', outFormat: 'hex' };
+    var opt = Object.assign(defaults, options);
```
Comment (zamentur) [inline delete.html]: For information: Not compatible with IE11.

**Comment by zamentur** [comment]
If you want set up the dynette quickly https://github.com/YunoHost-Apps/dynette_ynh

**Comment by Psycojoker** [comment]
@zamentur it would be interesting to add this information in the README

**Comment by alexAubin** [comment]
TODO : Check if we can avoid including an external hash compute code, use https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest instead ?

**Comment by Psycojoker** [comment]
If we want to go top notch super ubber crypto apparently this is the hashing algorithm we should uses instead of bcrypt which is not "the cool kid in town" anymore https://en.wikipedia.org/wiki/PBKDF2

**Comment by Psycojoker** [comment]
I haven't made a deep review but I think that this is good enough so it's an approval from me.

**Comment by alexAubin** [comment]
Successfully deployed on dyndns.yunohost.org ! You can set the recovery password using  https://github.com/YunoHost/yunohost/pull/335 then delete the domain from https://dyndns.yunohost.org/delete !

**Pull request merged**